File: tests/python/test_openeye_interop.py
"""Tests for OpenEye fingerprint interop."""

import pytest


def test_openeye_import_matches_openeye_tanimoto(aspirin_mol, ethanol_mol):
    from openeye import oegraphsim
    import oefp

    oe_a = oegraphsim.OEFingerPrint()
    oe_b = oegraphsim.OEFingerPrint()
    assert oegraphsim.OEMakeCircularFP(oe_a, aspirin_mol)
    assert oegraphsim.OEMakeCircularFP(oe_b, ethanol_mol)

    fp_a = oefp.from_openeye_fingerprint(oe_a)
    fp_b = oefp.from_openeye_fingerprint(oe_b)

    assert oefp.compare(fp_a, fp_b, oefp.Metric.tanimoto()) == pytest.approx(
        oegraphsim.OETanimoto(oe_a, oe_b)
    )


def test_openeye_export_roundtrip(ethanol_mol):
    from openeye import oegraphsim
    import oefp

    original = oegraphsim.OEFingerPrint()
    assert oegraphsim.OEMakeCircularFP(original, ethanol_mol)

    fp = oefp.from_openeye_fingerprint(original)
    exported = oefp.to_openeye_fingerprint(fp)

    assert oegraphsim.OEIsSameFPType(original, exported)
    assert oegraphsim.OETanimoto(original, exported) == pytest.approx(1.0)


def test_openeye_import_exposes_numeric_type_metadata(ethanol_mol):
    from openeye import oegraphsim
    import oefp

    original = oegraphsim.OEFingerPrint()
    assert oegraphsim.OEMakeCircularFP(original, ethanol_mol)
    assert original.GetFPTypeBase() is not None

    fp = oefp.from_openeye_fingerprint(original)
    spec = fp._native.Spec()

    assert spec.has_source_type_id
    assert spec.source_type_id == original.GetFPTypeBase().GetFPType()


def test_openeye_export_uses_numeric_type_metadata_when_string_empty(ethanol_mol):
    from openeye import oegraphsim
    import oefp

    original = oegraphsim.OEFingerPrint()
    assert oegraphsim.OEMakeCircularFP(original, ethanol_mol)

    fp = oefp.from_openeye_fingerprint(original)
    fp._native.Spec().source_type = ""

    exported = oefp.to_openeye_fingerprint(fp)

    assert oegraphsim.OEIsSameFPType(original, exported)
    assert oegraphsim.OETanimoto(original, exported) == pytest.approx(1.0)


def test_openeye_export_rejects_invalid_string_before_numeric_fallback(ethanol_mol):
    import oefp
    from openeye import oegraphsim

    original = oegraphsim.OEFingerPrint()
    assert oegraphsim.OEMakeCircularFP(original, ethanol_mol)

    fp = oefp.from_openeye_fingerprint(original)
    fp._native.Spec().source_type = "not-a-valid-openeye-fingerprint-type"

    with pytest.raises(ValueError, match="could not be resolved"):
        oefp.to_openeye_fingerprint(fp)

File: tests/cpp/test_openeye_interop.cpp
#include <gtest/gtest.h>

#include "oefp/compare.h"
#include "oefp/metric.h"
#include "oefp/openeye.h"

#include <oechem.h>
#include <oegraphsim.h>

#include <stdexcept>

namespace OEFP {
namespace test {

TEST(OpenEyeInteropTest, ImportsGeneratedFingerprintAndMatchesOpenEyeTanimoto) {
    OEChem::OEGraphMol mol_a;
    OEChem::OEGraphMol mol_b;
    ASSERT_TRUE(OEChem::OESmilesToMol(mol_a, "c1ccccc1"));
    ASSERT_TRUE(OEChem::OESmilesToMol(mol_b, "c1ccc(O)cc1"));

    OEGraphSim::OEFingerPrint oe_a;
    OEGraphSim::OEFingerPrint oe_b;
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(oe_a, mol_a));
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(oe_b, mol_b));

    const auto fp_a = FromOEFingerPrint(oe_a);
    const auto fp_b = FromOEFingerPrint(oe_b);

    EXPECT_EQ(fp_a.SizeBits(), oe_a.GetSize());
    EXPECT_EQ(fp_b.SizeBits(), oe_b.GetSize());
    EXPECT_NEAR(Compare(fp_a, fp_b, Metric::Tanimoto()), OEGraphSim::OETanimoto(oe_a, oe_b), 1.0e-7);
}

TEST(OpenEyeInteropTest, ImportsNumericFingerprintTypeMetadata) {
    OEChem::OEGraphMol mol;
    ASSERT_TRUE(OEChem::OESmilesToMol(mol, "CCO"));

    OEGraphSim::OEFingerPrint original;
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(original, mol));
    ASSERT_NE(original.GetFPTypeBase(), nullptr);

    const auto imported = FromOEFingerPrint(original);

    EXPECT_TRUE(imported.Spec().has_source_type_id);
    EXPECT_EQ(imported.Spec().source_type_id, original.GetFPTypeBase()->GetFPType());
}

TEST(OpenEyeInteropTest, ExportCanResolveTypeFromNumericMetadataFallback) {
    OEChem::OEGraphMol mol;
    ASSERT_TRUE(OEChem::OESmilesToMol(mol, "CCO"));

    OEGraphSim::OEFingerPrint original;
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(original, mol));
    const auto imported = FromOEFingerPrint(original);

    auto spec = imported.Spec();
    spec.source_type.clear();
    const OEFP fallback(spec, imported.Words());
    const auto exported = ToOEFingerPrint(fallback);

    EXPECT_TRUE(OEGraphSim::OEIsSameFPType(original, exported));
    EXPECT_NEAR(OEGraphSim::OETanimoto(original, exported), 1.0, 1.0e-7);
}

TEST(OpenEyeInteropTest, ExportRejectsInvalidStringTypeBeforeNumericFallback) {
    OEChem::OEGraphMol mol;
    ASSERT_TRUE(OEChem::OESmilesToMol(mol, "CCO"));

    OEGraphSim::OEFingerPrint original;
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(original, mol));
    const auto imported = FromOEFingerPrint(original);

    auto spec = imported.Spec();
    spec.source_type = "not-a-valid-openeye-fingerprint-type";
    const OEFP fallback(spec, imported.Words());

    EXPECT_THROW(ToOEFingerPrint(fallback), std::invalid_argument);
}

TEST(OpenEyeInteropTest, ExportRoundTripsGeneratedFingerprintBits) {
    OEChem::OEGraphMol mol;
    ASSERT_TRUE(OEChem::OESmilesToMol(mol, "CCO"));

    OEGraphSim::OEFingerPrint original;
    ASSERT_TRUE(OEGraphSim::OEMakeCircularFP(original, mol));

    const auto imported = FromOEFingerPrint(original);
    const auto exported = ToOEFingerPrint(imported);

    EXPECT_EQ(exported.GetSize(), original.GetSize());
    EXPECT_EQ(exported.CountBits(), original.CountBits());
    EXPECT_TRUE(OEGraphSim::OEIsSameFPType(original, exported));
    EXPECT_NEAR(OEGraphSim::OETanimoto(original, exported), 1.0, 1.0e-7);
}

TEST(OpenEyeInteropTest, ExportRejectsNonOpenEyeFingerprintSpec) {
    FingerprintSpec spec;
    spec.size_bits = 64;
    spec.value_type = FingerprintValueType::Binary;
    spec.source_name = "unit-test";
    spec.source_type = "manual";

    EXPECT_THROW(ToOEFingerPrint(OEFP(spec)), std::invalid_argument);
}

} // namespace test
} // namespace OEFP
